crates/client/rpc/src/versions/user/v0_8_0/methods/read/get_storage_proof.rs

```diff
@@ -0,0 +1,186 @@
+        .or_internal_server_error("Getting root hash of trie")?;
+
+    let proof = storage
+        .get_multi_proof(identifier, keys.iter().map(|k| &k.as_bitslice()[5..]))
+        .map_err(|err| anyhow::anyhow!("{err:#}"))
+        .or_internal_server_error("Error while making storage multiproof")?;
+
+    // convert the bonsai-trie type to the rpc DTO
+    let converted_proof = proof
+        .0
+        .into_iter()
+        .map(|(node_hash, n)| {
+            let node = match n {
+                mc_db::ProofNode::Binary { left, right } => MerkleNode::Binary { left, right },
+                mc_db::ProofNode::Edge { child, path } => {
+                    MerkleNode::Edge { child, path: path_to_felt(&path), length: path.len() }
+                }
+            };
+            NodeHashToNodeMappingItem { node_hash, node }
+        })
+        .collect();
+
+    Ok((root_hash, converted_proof))
+}
+
+pub fn get_storage_proof(
+    starknet: &Starknet,
+    block_id: BlockId,
+    class_hashes: Option<Vec<Felt>>,
+    contract_addresses: Option<Vec<Felt>>,
+    contracts_storage_keys: Option<Vec<ContractStorageKeysItem>>,
+) -> RpcResult<GetStorageProofResult> {
+    // Pending block does not have a state root, so always fallbacck to latest.
+    let block_id = match block_id {
+        BlockId::Tag(BlockTag::Pending) => BlockId::Tag(BlockTag::Latest),
+        block_id => block_id,
+    };
+
+    let block_n = starknet
+        .backend
+        .get_block_n(&block_id)
+        .or_internal_server_error("Resolving block number")?
+        .ok_or(StarknetRpcApiError::NoBlocks)?;
+
+    let latest = starknet.backend.get_latest_block_n().or_internal_server_error("Getting latest block in db")?;
+    if latest.is_none() || latest.is_some_and(|latest| block_n > latest) {
+        return Err(StarknetRpcApiError::BlockNotFound.into());
+    }
+    let block_hash = starknet
+        .backend
+        .get_block_hash(&block_id)
+        .or_internal_server_error("Resolving block hash")?
+        .ok_or(StarknetRpcApiError::NoBlocks)?;
+
+    let class_hashes = class_hashes.unwrap_or_default();
+    let contract_addresses = contract_addresses.unwrap_or_default();
+    let contracts_storage_keys = contracts_storage_keys.unwrap_or_default();
+
+    // Check limits.
+
+    let proof_keys = class_hashes.len()
+        + contract_addresses.len()
+        + contracts_storage_keys.iter().map(|v| v.storage_keys.len()).sum::<usize>();
+    if proof_keys > starknet.storage_proof_config.max_keys {
+        return Err(StarknetRpcApiError::ProofLimitExceeded {
+            kind: StorageProofLimit::MaxKeys,
+            limit: starknet.storage_proof_config.max_keys,
+            got: proof_keys,
+        }
+        .into());
+    }
+
+    let n_tries = (!class_hashes.is_empty() as usize)
+        + (!contract_addresses.is_empty() as usize)
+        + contracts_storage_keys.iter().map(|keys| (!keys.storage_keys.is_empty() as usize)).sum::<usize>();
```

Comment:
:warning: This should be `saturating_add` as we are dealing with user-controlled input!